probe-rs/src/probe/stlink/mod.rs

```diff
@@ -775,6 +786,262 @@ impl<D: StLinkUsb> STLink<D> {
+                apsel,
+            ],
+            &[],
+            &mut receive_buffer,
+            TIMEOUT,
+        )?;
+
+        // get last RW status
+        self.get_last_rw_status()?;
+
+        let words: Vec<u32> = receive_buffer
+            .chunks_exact(4)
+            .map(|chunk| u32::from_le_bytes(chunk.try_into().unwrap()))
+            .collect();
+
+        Ok(words)
+    }
+
+    fn read_mem_8bit(
+        &mut self,
+        address: u32,
+        length: u16,
+        apsel: u8,
+    ) -> Result<Vec<u8>, DebugProbeError> {
+        log::debug!("Read mem 8 bit, address={:08x}, length={}", address, length);
+        let mut receive_buffer = vec![0u8; length as usize];
+
+        self.device.write(
+            &[
+                commands::JTAG_COMMAND,
+                commands::JTAG_READMEM_8BIT,
+                address as u8,
+                (address >> 8) as u8,
+                (address >> 16) as u8,
+                (address >> 24) as u8,
+                length as u8,
+                (length >> 8) as u8,
+                apsel,
+            ],
+            &[],
+            &mut receive_buffer,
+            TIMEOUT,
+        )?;
+
+        // get last RW status
+        self.get_last_rw_status()?;
+
+        Ok(receive_buffer)
+    }
+
+    fn write_mem_32bit(
+        &mut self,
+        address: u32,
+        data: &[u32],
+        apsel: u8,
+    ) -> Result<(), DebugProbeError> {
+        log::trace!("write_mem_32bit");
+        let length = data.len();
+
+        // Maximum supported read length is 2^16 bytes
+        assert!(
+            length < (u16::MAX / 4) as usize,
+            "Maximum write length for STLink is 16'384 words"
+        );
+
+        if address % 4 != 0 {
+            todo!("Should return an error here");
+        }
+
+        let byte_length = length * 4;
+
+        let mut tx_buffer = vec![0u8; byte_length];
+
+        let mut offset = 0;
+
+        for word in data {
+            tx_buffer
+                .gwrite(word, &mut offset)
+                .expect("Failed to write into tx_buffer");
+        }
+
+        self.device.write(
+            &[
+                commands::JTAG_COMMAND,
+                commands::JTAG_WRITEMEM_32BIT,
+                address as u8,
+                (address >> 8) as u8,
+                (address >> 16) as u8,
+                (address >> 24) as u8,
+                byte_length as u8,
+                (byte_length >> 8) as u8,
+                apsel,
+            ],
+            &tx_buffer,
+            &mut [],
+            TIMEOUT,
+        )?;
+
+        // get last RW status
+        self.get_last_rw_status()?;
+
+        Ok(())
+    }
+
+    fn write_mem_8bit(
+        &mut self,
+        address: u32,
+        data: &[u8],
+        apsel: u8,
+    ) -> Result<(), DebugProbeError> {
+        log::trace!("write_mem_8bit");
+        let byte_length = data.len();
+
+        assert!(byte_length <= 64); // 64 byte for stlinkv2, 512 for stlinkv3
```

Comment:
This should really be a proper error instead of a panic.

Also, I am pretty sure this will panic for v3?

```suggestion
        // 64 bytes for stlinkv2, 512 bytes for stlinkv3.
        assert!(byte_length <= 64);
```

Reply:
Good catch, would work fine for v3, just slower than necessary.